File: spatial-index/src/main/java/lva/spatialindex/Entry.java
package lva.spatialindex;

import java.awt.*;

/**
 * @author vlitvinenko
 */
class Entry {
    private static int SIZE_OF_X = 4;
    private static int SIZE_OF_Y = 4;
    private static int SIZE_OF_WIDTH = 4;
    private static int SIZE_OF_HEIGHT = 4;
    private static int SIZE_OF_CHILD = 8;

    static int SIZE = SIZE_OF_X + SIZE_OF_Y + SIZE_OF_WIDTH + SIZE_OF_HEIGHT + SIZE_OF_CHILD;

    private final Storage<Node> storage;
    final long childOffset;
    Rectangle mbr; // TODO: think about to make final

    Entry(Storage<Node> storage, Rectangle mbr, long childOffset) {
        this.storage = storage;
        this.mbr = mbr;
        this.childOffset = childOffset;
    }

    public Node loadNode() {
        return childOffset >= 0 ? storage.get(childOffset) : null;
    }

    long getChildOffset() {
        return childOffset;
    }

    static Entry of(Storage<Node> storage, Rectangle mbr, long nodeOffset) {
        return new Entry(storage, mbr, nodeOffset);
    }
}

File: spatial-index/src/main/kotlin/lva/spatialindex/index/Entry.kt
package lva.spatialindex.index

import arrow.core.Either
import lva.spatialindex.storage.Storage
import java.awt.Rectangle


/**
 * @author vlitvinenko
 */
internal class Entry(private val storage: Storage<Node>, mbr: Rectangle, childOffset: Long) {
    private data class Body(var mbr: Rectangle, val childOffset: Long)
    private val body = Body(mbr, childOffset)

    var mbr by body::mbr
    val childOffset by body::childOffset
    val childNode get() = data().swap().orNull()
    val isLeaf get() = childOffset < 0

    fun data(): Either<Node, Long> = if (childOffset >= 0)
        Either.Left(storage.read(childOffset)) else Either.Right(childOffset)

    override fun equals(other: Any?) =
        if (other is Entry) body == other.body else false

    override fun hashCode() =
        body.hashCode()

    companion object {
        const val SIZE = 24
    }
}

internal fun List<Entry>.union(): Rectangle =
    asSequence().map { it.mbr }
        .reduceOrNull { acc, r -> acc.union(r) } ?: Rectangle()

internal fun List<Entry>.margin(): Int =
    sumOf { it.mbr.margin }
